internal/pkg/cli/deploy/env.go

```diff
@@ -0,0 +1,68 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package deploy
+
+import (
+	"fmt"
+
+	"github.com/aws/copilot-cli/internal/pkg/aws/s3"
+	"github.com/aws/copilot-cli/internal/pkg/config"
+	"github.com/aws/copilot-cli/internal/pkg/deploy/cloudformation/stack"
+	"github.com/aws/copilot-cli/internal/pkg/manifest"
+	"github.com/aws/copilot-cli/internal/pkg/term/log"
+)
+
+type appResourcesGetter interface {
+	GetAppResourcesByRegion(app *config.Application, region string) (*stack.AppRegionalResources, error)
+	GetRegionalAppResources(app *config.Application) ([]*stack.AppRegionalResources, error)
+}
+
+type envDeployer struct {
+	app *config.Application
+	env *config.Environment
+
+	mft manifest.Environment
+
+	appCFN appResourcesGetter
+
+	// Dependencies to upload artifacts.
+	uploader customResourcesUploader
+	s3       uploader
+
+	// Cached variables.
+	appRegionalResources *stack.AppRegionalResources
+}
+
+// UploadArtifacts uploads the deployment artifacts for the environment.
+func (d *envDeployer) UploadArtifacts() (map[string]string, error) {
+	envRegion := d.env.Region
+
+	resources, err := d.getAppRegionalResources()
+	if err != nil {
+		return nil, err
+	}
+	if resources.S3Bucket == "" {
+		log.Errorf("Cannot find the S3 artifact bucket in %s region created by app %s. The S3 bucket is necessary for many future operations. For example, when you need addons to your services.", envRegion, d.app.Name)
```

Comment:
Should we break this up into two or three lines, for width reasons?

Reply:
Why do we need this log msg? I feel it is not a common error unless they remove the bucket manually?

Reply:
The log message was moved over from `cli/env_init.go` but sure we can remove it!